### PR TITLE
Fix S3 Event not deserialized properly

### LINKED ISSUE #1940: S3Event not properly deserialized
### Expected Behavior

`getRecords()` to have values

### Actual Behaviour

`getRecords()` is null

### Steps To Reproduce

_No response_

### Environment Information

_No response_

### Example Application

_No response_

### Version

4.1.5

### PR DESCRIPTION
**Pull request by driverpt**
Fix #1940

### CODE REVIEW

**Review by sdelamo — COMMENTED** [review]
For S3, AWS SDK uses this workaround: 

com.amazonaws.services.lambda.runtime.serialization.events.serializers.S3EventSerializer. 

I am not sure the correct thing to do is to create a Mixin. Maybe we should try to port the contents of that class to a [custom Micronaut Serde serializer](https://micronaut-projects.github.io/micronaut-serialization/latest/guide/#serdes).

**Review by sdelamo — COMMENTED** [review]
I added `Serde` for Object Entity and for `ResponseElementsEntity`. `S3EventNotification` is working now for `S3EventNotification` but not for `S3Event` yet.

File: aws-lambda-events-serde/src/main/java/io/micronaut/aws/lambda/events/serde/ResponseElementsEntitySerde.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ */
+@Internal
+@Singleton
+public class ResponseElementsEntitySerde implements Serde<S3EventNotification.ResponseElementsEntity> {
+    private static final String X_AMZ_ID_2 = "x-amz-id-2";
+    private static final String X_AMZ_REQUEST_ID = "x-amz-request-id";
+
+    @Override
+    public @Nullable S3EventNotification.ResponseElementsEntity deserialize(@NonNull Decoder decoder, @NonNull DecoderContext context, @NonNull Argument<? super S3EventNotification.ResponseElementsEntity> type) throws IOException {
+        JsonNode node = decoder.decodeNode();
```
Comment (sdelamo) [inline aws-lambda-events-serde/src/main/java/io/micronaut/aws/lambda/events/serde/ResponseElementsEntitySerde.java:42]: @yawkat is this the correct way to use the `Decoder` API?

File: aws-lambda-events-serde/src/main/java/io/micronaut/aws/lambda/events/serde/ResponseElementsEntitySerde.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        JsonNode xAmzId2Node = node.get(X_AMZ_ID_2);
+        String xAmzId2 = xAmzId2Node != null ? xAmzId2Node.getStringValue() : null;
+        JsonNode xAmzRequestIdNode = node.get(X_AMZ_REQUEST_ID);
+        String xAmzRequestId = xAmzRequestIdNode != null ? xAmzRequestIdNode.getStringValue() : null;
+        return new S3EventNotification.ResponseElementsEntity(xAmzId2, xAmzRequestId);
+    }
+
+    @Override
+    public void serialize(@NonNull Encoder encoder, @NonNull EncoderContext context, @NonNull Argument<? extends S3EventNotification.ResponseElementsEntity> type, S3EventNotification.@NonNull ResponseElementsEntity value) throws IOException {
+        encoder.encodeObject(type);
```
Comment (sdelamo) [inline aws-lambda-events-serde/src/main/java/io/micronaut/aws/lambda/events/serde/ResponseElementsEntitySerde.java:52]: @yawkat is this the correct way to use the `Encoder` API?

File: aws-lambda-events-serde/src/main/java/io/micronaut/aws/lambda/events/serde/S3ObjectEntitySerde.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+public class S3ObjectEntitySerde implements Serde<S3EventNotification.S3ObjectEntity> {
+    private final String KEY = "key";
+    private final String SIZE = "size";
+    private final String ETAG = "eTag";
+    private final String VERSION_ID = "versionId";
+    private final String SEQUENCER = "sequencer";
+
+    @Override
+    public @Nullable S3EventNotification.S3ObjectEntity deserialize(@NonNull Decoder decoder, @NonNull DecoderContext context, @NonNull Argument<? super S3EventNotification.S3ObjectEntity> type) throws IOException {
+        JsonNode node = decoder.decodeNode();
```
Comment (sdelamo) [inline aws-lambda-events-serde/src/main/java/io/micronaut/aws/lambda/events/serde/S3ObjectEntitySerde.java:45]: @yawkat is this the correct way to use the `Decoder` API?

File: aws-lambda-events-serde/src/main/java/io/micronaut/aws/lambda/events/serde/S3ObjectEntitySerde.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+        JsonNode sequencerNode = node.get(SEQUENCER);
+        String sequencer = sequencerNode != null ? sequencerNode.getStringValue() : null;
+
+        return new  S3EventNotification.S3ObjectEntity(key, size, eTag, versionId, sequencer);
+    }
+
+    @Override
+    public void serialize(@NonNull Encoder encoder, @NonNull EncoderContext context, @NonNull Argument<? extends S3EventNotification.S3ObjectEntity> type, S3EventNotification.@NonNull S3ObjectEntity value) throws IOException {
+        encoder.encodeObject(type);
```
Comment (sdelamo) [inline aws-lambda-events-serde/src/main/java/io/micronaut/aws/lambda/events/serde/S3ObjectEntitySerde.java:67]: @yawkat is this the correct way to use the `Encoder` API?

**Comment by yawkat** [comment]
ive implemented it differently.

btw is there a reason why the serdeimports in that package are split among 100 classes? it makes it a bit confusing imo

**Comment by sdelamo** [comment]
> ive implemented it differently.
> 
> btw is there a reason why the serdeimports in that package are split among 100 classes? it makes it a bit confusing imo

The reason is to have to group them for the class they refer to.

**Review by sdelamo — COMMENTED** [review]
@yawkat could you check it to make it work for S3Event as well: 

```
@PendingFeature
void "S3Event can be deserialized with s3-put"() {
```

see: 

```
@PendingFeature
void "S3Event can be serialized with s3-event"() {
```

**Review by sdelamo — APPROVED** [review]

**Comment by brucej72** [comment]
@sdelamo has this fix gone out into a micronaut release yet? I have 4.3.0 installed and I am still seeing this problem

**Comment by sdelamo** [comment]
Thanks for reporting @brucej72  Micronaut Framework 4.3.5 should solve this issue.

**Pull request merged**